jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -589,56 +628,397 @@ else if (StringUtil.asciiStartsWithIgnoreCase(contextPath, "root-"))
         {
             int dash = contextPath.indexOf('-');
             String virtual = contextPath.substring(dash + 1);
-            context.setVirtualHosts(Arrays.asList(virtual.split(",")));
+            contextHandler.setVirtualHosts(Arrays.asList(virtual.split(",")));
             contextPath = "/";
         }
 
         // Ensure "/" is Prepended to all context paths.
         if (contextPath.charAt(0) != '/')
             contextPath = "/" + contextPath;
 
-        // Set the display name and context Path
-        context.setDisplayName(basename);
-        context.setContextPath(contextPath);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with displayName: {}", contextHandler, basename);
+        contextHandler.setDisplayName(basename);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with contextPath: {}", contextHandler, contextPath);
+        contextHandler.setContextPath(contextPath);
+    }
+
+    /**
+     * Apply the main deployable heuristics referenced in the main javadoc
+     * for this class.
+     *
+     * @param unit the unit of deployment to interrogate.
+     * @return the main deployable path
+     */
+    @Override
+    protected Path getMainDeploymentPath(Unit unit)
```

Comment:
It is a bit strange that the logic that knows how to group different paths into a single unit is in the internal `DeploymentUnits`, whilst the logic to pick which of them is the main deployment path is here, in an entirely different place.     Surely both bits of logic are related and should be in the same class

Reply:
Picking the main path seems more like a ContextProvider concern.
The Unit and DeploymentUnits is go-between from the Scanner events to the ContextProvider.

Reply:
I don't think so, because even working out if a Unit is ADDED, CHANGED or REMOVED needs knowledge of deployable files, so adding a priority of deployable files is just a tiny step